File: day-1/part-2.ts
import { readFileSync } from "fs";
import { join } from "path";

// const sample = [100, 101, 105, 106, 103, 104, 106, 108, 112, 123]; // expected 8

const getSlidingWindowsSums = (data: number[], windowSize = 3) => {
    const measurements = [...data];

    const windowSums = [];

    for(let i=0; i<data.length; i++){
        const window = measurements.slice(i, i + windowSize);

        if(window.length === windowSize){
            const sum = window.reduce((sum, m) => sum + m, 0);
            windowSums.push(sum);
        }
    }

    return windowSums;
};

const getMeasurementsIncrease = (measurements: number[]) => {
    if(measurements.length === 0) throw new Error("Measurements are empty.");

    let increase = 0;
    let prev = measurements[0];
    
    for(const measurement of measurements){
        if(measurement > prev) increase++;
        prev = measurement;
    }
    
    return increase;
};

const loadInput = (file: string) => {
    const data = readFileSync(file, { encoding: "utf-8" });

    const input = data.split("\n").map(row => parseInt(row));

    return input;
};


(() => {
    const inputFile = join(__dirname, "input");

    const input = loadInput(inputFile);
    const slidingWindowsSums = getSlidingWindowsSums(input, 3);
    const result = getMeasurementsIncrease(slidingWindowsSums);

    console.log(result);
})();

File: day-1/part-2.py
import os

def load_input(file):
    with open(file, "r") as lines:
        return [int(line) for line in lines]

def get_window_sums(data, window_size=3):
    sums = []

    for i in range(len(data)):
        window = data[i:i+window_size]
        if len(window) is window_size:
            sums.append(sum(window))
    
    return sums
            
def get_increases(measurements):
    count = 0

    for i in range(1, len(measurements)):
        if measurements[i] > measurements[i-1]:
            count += 1
    
    return count

def main():
    file_name = os.path.join(os.path.dirname(__file__), "input")

    input = load_input(file_name)
    window_sums = get_window_sums(input, 3)
    count = get_increases(window_sums)

    print(count)

if __name__ == '__main__':
    main()
    